File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.py
"""
0. 입력
- N: 정점, M, 간선

1. 아이디어
- 인접리스트 dfs
    - 방문 노드 check
- dfs가 끊기면 +1

2. 시간복잡도
인접리스트 dfs => O(V+E)

3. 자료구조
- int[][] => 인접리스트
"""

import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)
N, M = map(int, input().split(" "))

graph = [[] for _ in range(N+1)]
chk = [False] * (N+1)
for _ in range(M):
    a, b = map(int, input().split(" "))
    graph[a].append(b)
    graph[b].append(a)

def dfs(x):
    for v in graph[x]:
        if chk[v] == False:
            chk[v] = True
            dfs(v)

cnt = 0
for i in range(1, N+1):
    if chk[i] == False:
        chk[i] = True
        cnt += 1
        dfs(i)

print(cnt)

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

public class Main {

    static int[][] graph;
    static int count;
    static boolean[] check;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int[] input = Arrays.stream(br.readLine().split(" "))
                .mapToInt(Integer::parseInt)
                .toArray();

        int n = input[0];
        int m = input[1];

        graph = new int[n + 1][n + 1];
        check = new boolean[n + 1];

        for (int i = 0; i < m; i++) {
            int[] x = Arrays.stream(br.readLine().split(" "))
                    .mapToInt(Integer::parseInt)
                    .toArray();
            int a = x[0];
            int b = x[1];
            graph[a][b] = 1;
            graph[b][a] = 1;
        }

        for (int i = 1; i < n + 1; i++) {
            if (!check[i]) {
                count++;
//                dfs(i);
                bfs(i);
            }
        }
        System.out.println(count);
    }

    static Queue<Integer> q;
    public static void bfs(int node) {
        q = new LinkedList<>();
        q.offer(node);
        check[node] = true;

        while(!q.isEmpty()) {
            int now = q.poll();
            for (int i = 1; i < graph[now].length; i++) {
                if (graph[now][i] == 1 && !check[i]) {
                    check[i] = true;
                    q.add(i);
                }
            }
        }
    }

    public static void dfs(int node) {
        check[node] = true;

        for (int i = 1; i < graph[node].length; i++) {
            if (graph[node][i] == 1 && !check[i]) {
                dfs(i);
            }
        }
    }
}
